nav2_collision_monitor/src/polygon_source.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2023 Pixel Robotics GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/polygon_source.hpp"
+
+#include <functional>
+
+#include "geometry_msgs/msg/polygon_stamped.hpp"
+
+#include "nav2_util/node_utils.hpp"
+#include "nav2_util/robot_utils.hpp"
+
+using rcl_interfaces::msg::ParameterType;
+
+namespace nav2_collision_monitor
+{
+
+PolygonSource::PolygonSource(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string & source_name,
+  const std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+  const std::string & base_frame_id,
+  const std::string & global_frame_id,
+  const tf2::Duration & transform_tolerance,
+  const rclcpp::Duration & source_timeout,
+  const bool base_shift_correction)
+: Source(
+    node, source_name, tf_buffer, base_frame_id, global_frame_id,
+    transform_tolerance, source_timeout, base_shift_correction),
+  data_(nullptr)
+{
+}
+
+PolygonSource::~PolygonSource()
+{
+  data_sub_.reset();
+}
+
+void PolygonSource::configure()
+{
+  Source::configure();
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  std::string source_topic;
+
+  getParameters(source_topic);
+
+  rclcpp::QoS qos = rclcpp::SensorDataQoS();  // set to default
+  data_sub_ = node->create_subscription<nav2_msgs::msg::PolygonsArray>(
+    source_topic, qos,
+    std::bind(&PolygonSource::dataCallback, this, std::placeholders::_1));
+}
+
+void PolygonSource::getData(
+  const rclcpp::Time & curr_time,
+  std::vector<Point> & data) const
+{
+  // Ignore data from the source if it is not being published yet or
+  // not published for a long time
+  if (data_ == nullptr || data_->polygons.empty()) {
+    return;
+  }
+  // get the oldest time stamp from the polygon array
+  rclcpp::Time oldest_stamp = rclcpp::Time(data_->polygons[0].header.stamp);
+  for (const auto & polygon : data_->polygons) {
+    if (rclcpp::Time(polygon.header.stamp) < oldest_stamp) {
+      oldest_stamp = rclcpp::Time(polygon.header.stamp);
+    }
+  }
+  if (!sourceValid(oldest_stamp, curr_time)) {
+    return;
+  }
+
+  geometry_msgs::msg::TransformStamped tf;
+  if (base_shift_correction_) {
+    // Obtaining the transform to get data from source frame and time where it was received
+    // to the base frame and current time
+    if (
+      !nav2_util::getTransform(
+        data_->header.frame_id, data_->header.stamp,
```

Comment:
Since each polygon in the array might be in its own frame, I believe, we need to get transform per each polygon it the message instead of using main headers' frame_id

Reply:
good point, changed